### ISSUE TITLE
ECS Agent filling up root drive on ECS-optimized instances

### ISSUE THREAD

**Issue by cwndrws** [aws/amazon-ecs-agent#764]
Yesterday, I started receiving alerts that some of my ECS container instances' root drives were filling up. I have a block device mapping configured for the `/dev/xvdcz` device, which Docker device mapper uses. That device looked normal, but `/dev/xvda1`, which is the root drive on these boxes was receiving a lot of writes. Using Datadog, I was able to determine the only thing writing at higher rate than normal was the ecs-agent container. 

![screen shot 2017-04-12 at 10 20 24 am](https://cloud.githubusercontent.com/assets/1382553/24965478/bf58ba70-1f69-11e7-8d99-e900fc37c4c1.png)

Using `du`, I narrowed the disk usage increase to `/var/lib/docker` but wasn't able to find out what the container was trying to write to. I am not certain what changed to cause this increase in disk usage. As a temporary fix, I am now mounting a larger EBS volume (100GB) to the root drive, but I would like to go back to the default 8GB root drive if possible. What can I do to further debug this disk usage problem.

**Comment by cabbruzzese** [comment]
@cwndrws I'm sorry that you are experiencing this issue. I'd like to gather some more information to help pinpoint what is going on here. 

It would be helpful to see the ECS agent version, the agent logs, docker version, and the docker debugging info.

* These [Instructions](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-agent-update.html#checking_agent_version) will help you find the agent version.
* Information on retrieving the instructions can be found in the following [documentation](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/troubleshooting.html#logs). Or you can collect them using the [ECS Logs Collector](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/troubleshooting.html#ecs-logs-collector).
* The docker version can be found with the command "sudo docker version"
* The docker debbuging info can be found with the command "sudo docker info"

If you are not comfortable posting it publicly on GitHub, you can email it directly to me at abbrcale at amazon.com.

**Comment by cwndrws** [comment]
Here is my current setup:
Agent version: 1.14.1
Docker version: 1.12.6
I wouldn't have logs from the problem anymore since I have rebuilt the spot fleet by which these boxes were created, so I could add larger root drives to them. I also saw the problem with an older version of the agent and docker. The older versions were something like:
Agent version: 1.13.x
Docker version: 1.11.x
My first fix was to upgrade my agent, after I did this it was fine for a few hours and then started writing a ton to disk again. What is the recommended root volume size for the ECS-optimized ami or any box running the ECS agent?

**Comment by cwndrws** [comment]
I just talked to a gentleman at the AWS booth at dockercon who suggested my problem was caused by containers not getting cleaned up fast enough and suggested I set the parameter container cleanup to be something like 30 minutes.

**Comment by cabbruzzese** [comment]
@cwndrws Did that suggestion help to remedy the issue?

Another potential cause is the way docker pulls images with Device Mapper as the storage driver. Data is saved to /var/lib/docker before it is moved to the thin pool.

**Comment by cwndrws** [comment]
@cabbruzzese Yes it did. I believe the solution I was given was the correct one. I don't believe it was caused by images being pulled, since these machines are running many containers that use the same image. The only images that would be pulled in the life time of these machines would be the ecs-agent image and the application image that is running on them. We do not update the application image often (maybe once or twice per week), so I don't believe the image pulling is the culprit here. I implemented the change to the `ECS_ENGINE_TASK_CLEANUP_WAIT_DURATION` setting and it seems to be doing fine now.
